### PR TITLE
JDK-8295865: Several issues with os::realloc

### PR DESCRIPTION
**Pull request by tstuefe**
There are several issues with os::realloc():

1) If realloc(3) fails, the original block will be left untouched. That is fine, and the caller may expect this and continue to use the old block, including handing it to os::free() later. But NMT has marked the original block as dead already, and subsequent os::free() or os::realloc() will trigger a false fatal block integrity error. Therefore, if realloc(3) fails, we need to revive the NMT header and re-account the original block before returning.

2) If handed in very large sizes, the size may overflow if the NMT header is added. The result would be that the VM reallocates to a much smaller buffer which would cause subsequent memory corruption if the caller were to use the buffer (same as JDK-8286519, but for realloc).

3) If os::realloc() enlarges a buffer, the newly added memory should be zapped with uinitBlockPad as we do for os::malloc(). Of course we only can do this if NMT is enabled, otherwise we won't know the original block size.

------------------------------

The patch fixes all three issues and adds hopefully thorough enough regression gtests for them. Remember that these gtests will run, as part of the jtreg gtest runners, in all NMT modes.

The largest diff hunk is inside os::realloc(). Note that I separated the two cases NMT=summary/detail and NMT=off, since it makes the coding easier to understand and we also run less code in the standard case of NMT==off.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295865](https://bugs.openjdk.org/browse/JDK-8295865): Several issues with os::realloc

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10857/head:pull/10857` \
`$ git checkout pull/10857`

Update a local copy of the PR: \
`$ git checkout pull/10857` \
`$ git pull https://git.openjdk.org/jdk pull/10857/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10857`

View PR using the GUI difftool: \
`$ git pr show -t 10857`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10857.diff">https://git.openjdk.org/jdk/pull/10857.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
32-bit errors unrelated. To be sure I built 32-bit manually on x86, and ran gtests with and without NMT, all is well.

**Comment by tstuefe** [comment]
Tested in SAP nightlies on all our platforms, no problems surfaced

File: src/hotspot/share/runtime/os.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/os.cpp:732]: I'm not familiar with NMT details - are `memflags` and `stack` guaranteed to be the same as the original allocation?

File: test/hotspot/gtest/testutils.hpp
Comment (dholmes-ora) [inline test/hotspot/gtest/testutils.hpp:57]: These aren't "convenience asserts" - should they be somewhere else? What does EXPECT_NULL actually do?

**Review by dholmes-ora — COMMENTED** [review]
This all appears to be quite reasonable.

I have a couple of queries.

Thanks.

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:732]: Not guaranteed, but it can happen.

`memflags` can theoretically be different, the memory would then just be re-booked to the new category. In practice, nobody does this.

`stack` can differ. Memory is accounted per { stack, flags } tupel in a hash table. Only matters for nmt "details" level though.

File: test/hotspot/gtest/testutils.hpp
Comment (tstuefe) [reply test/hotspot/gtest/testutils.hpp:57]: They extend the existing `ASSERT_xxx` and `EXPECT_xxx` macros we get from gtest (in gtest.h). I cannot add them to gtest, of course, they have to live within our gtest wrapper code. This is as good a place as any IMO. Note we do similar things with other test helpers, e.g. threadHelper.inline.hpp.

File: src/hotspot/share/runtime/os.cpp
Comment (dholmes-ora) [reply src/hotspot/share/runtime/os.cpp:732]: Don't you need to restore things exactly the way they were?

File: test/hotspot/gtest/testutils.hpp
Comment (tstuefe) [reply test/hotspot/gtest/testutils.hpp:57]: > I meant different place in the file, rather than under "convenience asserts" when they are not assertions. I remain ignorant as to what the behaviour of EXPECT_X(p) is actually meant to be ??

EXPECT_x and ASSERT_x mimic the assertion macros with the same name that the gtest suite exposes. This is gtest lingo. It is assumed here that the reader knows how gtest tests are written, in particular what ASSERT and EXPECT are and how they work.

I can make it more obvious with a clarifying comment.

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:732]: Shoot, you are right. memflags would be easy, but stack not so much. 

This may be difficult to fix. Note that the only negative effect this has is that in the very low probability case that realloc fails  *and* we run with NMT=detail, the original allocation is re-booked under the wrong stack.

I'll check how difficult this is to fix.

**Comment by dholmes-ora** [comment]
> @dholmes-ora what do you think? Should I revert to the earlier, simpler version?

Sorry @tstuefe I don't know enough about NMT to evaluate the two versions.

**Comment by tstuefe** [comment]
> > @dholmes-ora what do you think? Should I revert to the earlier, simpler version?
> 
> Sorry @tstuefe I don't know enough about NMT to evaluate the two versions.

@dholmes-ora: okay, then I vote to revert the patch back to its earlier, simpler form. I don't like the added complexity just to prevent a very rare and rather benign accounting error.

**Comment by tstuefe** [comment]
@dholmes-ora : anything else you want me to fix?

May I get a second review please?

**Comment by tstuefe** [comment]
@dholmes-ora : anything else you want me to fix? If not, could you please approve? Or do you have general objections to this patch?

To others: May I get a second review please?

**Review by dholmes-ora — APPROVED** [review]
I was waiting to see if anyone else chimed in on the "benign accounting error". I don't use NMT myself so don't know what impact it may have.

**Review by jdksjolen — APPROVED** [review]
Hi Thomas, thank you for your changes.

I've read through the code and explored some of NMT and I think that this is a good enhancement, so I will approve it. I believe that there is an elegant solution to the revival of NMT header+footers, but it can wait for a future RFE.

**Comment by tstuefe** [comment]
> Hi Thomas, thank you for your changes.
> 
> I've read through the code and explored some of NMT and I think that this is a good enhancement, so I will approve it.

Thank you!

> I believe that there is an elegant solution to the revival of NMT header+footers, but it can wait for a future RFE.

Yes, I have also plans. We should talk before walking in different directions.

Cheers, Thomas

**Pull request closed without merging**
